client/cookiejar.go

```diff
@@ -0,0 +1,234 @@
+func (cj *CookieJar) SetByHost(host []byte, cookies ...*fasthttp.Cookie) {
+	hostStr := utils.UnsafeString(host)
+
+	cj.mu.Lock()
+	defer cj.mu.Unlock()
+
+	if cj.hostCookies == nil {
+		cj.hostCookies = make(map[string][]*fasthttp.Cookie)
+	}
+
+	hostCookies, ok := cj.hostCookies[hostStr]
+	if !ok {
+		// If the key does not exist in the map, then we must make a copy for the key to avoid unsafe usage.
+		hostStr = string(host)
+	}
+
+	for _, cookie := range cookies {
+		c := searchCookieByKeyAndPath(cookie.Key(), cookie.Path(), hostCookies)
+		if c == nil {
+			// If the cookie does not exist in the slice, let's acquire new cookie and store it.
+			c = fasthttp.AcquireCookie()
+			hostCookies = append(hostCookies, c)
+		}
+		c.CopyTo(cookie) // override cookie properties
+	}
+	cj.hostCookies[hostStr] = hostCookies
+}
+
+// SetKeyValue sets a cookie by key and value for a specific host.
+//
+// This function prevents extra allocations by making repeated cookies
+// not being duplicated.
+func (cj *CookieJar) SetKeyValue(host, key, value string) {
+	cj.SetKeyValueBytes(host, utils.UnsafeBytes(key), utils.UnsafeBytes(value))
+}
+
+// SetKeyValueBytes sets a cookie by key and value for a specific host.
+//
+// This function prevents extra allocations by making repeated cookies
+// not being duplicated.
+func (cj *CookieJar) SetKeyValueBytes(host string, key, value []byte) {
+	c := fasthttp.AcquireCookie()
+	c.SetKeyBytes(key)
+	c.SetValueBytes(value)
+
+	cj.SetByHost(utils.UnsafeBytes(host), c)
+}
+
+// dumpCookiesToReq dumps the stored cookies to the request.
+func (cj *CookieJar) dumpCookiesToReq(req *fasthttp.Request) {
+	uri := req.URI()
+
+	cookies := cj.getByHostAndPath(uri.Host(), uri.Path())
+	for _, cookie := range cookies {
+		req.Header.SetCookieBytesKV(cookie.Key(), cookie.Value())
+	}
+}
+
+// parseCookiesFromResp parses the response cookies and stores them.
+func (cj *CookieJar) parseCookiesFromResp(host, path []byte, resp *fasthttp.Response) {
+	hostStr := utils.UnsafeString(host)
+
+	cj.mu.Lock()
+	defer cj.mu.Unlock()
+
+	if cj.hostCookies == nil {
+		cj.hostCookies = make(map[string][]*fasthttp.Cookie)
+	}
+	cookies, ok := cj.hostCookies[hostStr]
+	if !ok {
+		// If the key does not exist in the map then
+		// we must make a copy for the key to avoid unsafe usage.
+		hostStr = string(host)
+	}
+
+	now := time.Now()
+	resp.Header.VisitAllCookie(func(key, value []byte) {
+		isCreated := false
+		c := searchCookieByKeyAndPath(key, path, cookies)
+		if c == nil {
+			c, isCreated = fasthttp.AcquireCookie(), true
+		}
+
+		_ = c.ParseBytes(value) //nolint:errcheck // ignore error
+		if c.Expire().Equal(fasthttp.CookieExpireUnlimited) || c.Expire().After(now) {
+			cookies = append(cookies, c)
+		} else if isCreated {
+			fasthttp.ReleaseCookie(c)
+		}
+	})
+	cj.hostCookies[hostStr] = cookies
+}
+
+// Release releases all cookie values.
+func (cj *CookieJar) Release() {
+	cj.hostCookies = nil
```

Comment:
```go
	for _, v := range cj.hostCookies {
		for _, c := range v {
			fasthttp.ReleaseCookie(c)
		}
	}
```
why did you take this away? i think it would make sense to put the cookies back into the pool

Reply:
follow up ticket